crates/bootstrap_srv/src/space.rs

```diff
@@ -0,0 +1,228 @@
+use std::sync::{Arc, Mutex};
+
+use crate::store::*;
+
+/// A map of spaces.
+#[derive(Clone)]
+pub struct SpaceMap(Arc<Mutex<HashMap<bytes::Bytes, Space>>>);
+
+impl Default for SpaceMap {
+    fn default() -> Self {
+        Self(Arc::new(Mutex::new(HashMap::new())))
+    }
+}
+
+impl SpaceMap {
+    /// Read the content of a space.
+    pub fn read(&self, space: &bytes::Bytes) -> std::io::Result<Vec<u8>> {
+        // minimize outer mutex lock time
+        let space = self.0.lock().unwrap().get(space).cloned();
+
+        match space {
+            Some(space) => space.read(),
+            None => Ok(b"[]".to_vec()),
+        }
+    }
+
+    /// Update all the spaces stored in this map.
+    pub fn update_all(&self, max_entries: usize) {
+        // minimize outer mutex lock time
+        let all = self.0.lock().unwrap().keys().cloned().collect::<Vec<_>>();
+
+        for space in all {
+            self.update(max_entries, space, None);
+        }
+    }
+
+    /// Update a space, clearing out any expired infos and optionally
+    /// adding a new incoming info. If adding a new info, perform all
+    /// validation before calling this function.
+    pub fn update(
+        &self,
+        max_entries: usize,
+        space: bytes::Bytes,
+        new_info: Option<(crate::ParsedEntry, Option<StoreEntryRef>)>,
+    ) {
+        // minimize outer mutex lock time
+        let space: Space = {
+            use std::collections::hash_map::Entry;
+
+            let mut map = self.0.lock().unwrap();
+
+            if new_info.is_none() && !map.contains_key(&space) {
+                // we don't have this space, and we're not
+                // adding an info to it... nothing to do
+                return;
+            }
+
+            match map.entry(space) {
+                Entry::Occupied(e) => {
+                    // most other naive methods for clearing out dead spaces
+                    // result in either races or deadlock, or they require
+                    // us to hold the outer mutex lock for an unacceptably
+                    // long time. So, just doing this for now.
+                    if new_info.is_none()
+                        && e.get().readable.lock().unwrap().is_empty()
+                    {
+                        e.remove();
+                        return;
+                    }
+
+                    e.get().clone()
+                }
+                Entry::Vacant(e) => e.insert(Space::default()).clone(),
+            }
+        };
+
+        // do the actual update without the outer mutex locked
+        space.update(max_entries, new_info);
+    }
+}
+
+/// A concurrent single space.
+#[derive(Clone)]
+struct Space {
+    // Using a separate write lock instead of a RwLock, because
+    // we still want to allow reads while the write_lock is locked.
+    // The write function will lock the readable when it is done processing
+    // and quickly update the reader, then drop both locks (read first).
```

Comment:
I don't understand the difference between this behavior and RwLock, but as long as others do that's fine with me.

Reply:
I see so we need the write lock for a long time because the re-write code is processing and mutating the data in one go, rather than constructing a new list with only a read lock, then at the end getting the write lock and writing it all at once.

And there is no downside to allowing readers at the same time because a single entry will be mutated atomically, and so there's no risk of a reader getting some partially-written entry change? (forgive my ignorance)

Reply:
Since the infos are only stored on disk, the logic to update them means reading all the infos individually then doing any reorganization of the list of read references needed and then updating the specific reference list in memory that the read requests are cloning.

Since the read references themselves are not changing at all, it's that specific list in memory that readers are cloning that is okay for them to keep cloning. They will just get the older list (that may contain some recently expired infos, but that's okay).

Then that is locked and swapped out atomically, so thereafter read requests will get the updated list.

(no worries, understanding is what code review is all about! : )